tonic/src/transport/service/timeout.rs

```diff
@@ -0,0 +1,251 @@
+use http::Request;
+use http::{HeaderMap, HeaderValue};
+use pin_project::pin_project;
+use std::{fmt, pin::Pin};
+use std::{
+    future::Future,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::Sleep;
+use tower_service::Service;
+
+#[derive(Debug, Clone)]
+pub(crate) struct Timeout<S> {
+    inner: S,
+    server_timeout: Option<Duration>,
+}
+
+impl<S> Timeout<S> {
+    pub(crate) fn new(inner: S, server_timeout: Option<Duration>) -> Self {
+        Self {
+            inner,
+            server_timeout,
+        }
+    }
+}
+
+impl<S, ReqBody> Service<Request<ReqBody>> for Timeout<S>
+where
+    S: Service<Request<ReqBody>>,
+    S::Error: Into<crate::Error>,
+{
+    type Response = S::Response;
+    type Error = crate::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let client_timeout = try_parse_grpc_timeout(req.headers()).unwrap_or_else(|e| {
+            tracing::trace!("Error parsing `grpc-timeout` header {:?}", e);
+            None
+        });
+
+        // Use the shorter of the two durations, if either are set
+        let timeout_duration = match (client_timeout, self.server_timeout) {
+            (None, None) => None,
+            (Some(dur), None) => Some(dur),
+            (None, Some(dur)) => Some(dur),
+            (Some(header), Some(server)) => {
+                let shorter_duration = std::cmp::min(header, server);
+                Some(shorter_duration)
+            }
+        };
+
+        ResponseFuture {
+            inner: self.inner.call(req),
+            sleep: timeout_duration
+                .map(tokio::time::sleep)
+                .map(OptionPin::Some)
+                .unwrap_or(OptionPin::None),
+        }
+    }
+}
+
+#[pin_project]
+pub(crate) struct ResponseFuture<F> {
+    #[pin]
+    inner: F,
+    #[pin]
+    sleep: OptionPin<Sleep>,
+}
+
+impl<F, Res, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Res, E>>,
+    E: Into<crate::Error>,
+{
+    type Output = Result<Res, crate::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        if let Poll::Ready(result) = this.inner.poll(cx) {
+            return Poll::Ready(result.map_err(Into::into));
+        }
+
+        if let OptionPinProj::Some(sleep) = this.sleep.project() {
+            futures_util::ready!(sleep.poll(cx));
+            return Poll::Ready(Err(TimeoutExpired.into()));
+        }
+
+        Poll::Pending
+    }
+}
+
+#[pin_project(project = OptionPinProj)]
+enum OptionPin<T> {
+    Some(#[pin] T),
+    None,
+}
+
+const GRPC_TIMEOUT_HEADER: &str = "grpc-timeout";
```

Comment:
This should probably live in code that contains the list of all the other reserved headers?